File: src/main/java/team/zeromods/ancientmagic/client/MagicBookScreen.java
package team.zeromods.ancientmagic.client;

import api.ancientmagic.magic.MagicTypes;
import api.ancientmagic.mod.Constant;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import net.minecraft.ChatFormatting;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.gui.screens.inventory.InventoryScreen;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import team.zeromods.ancientmagic.init.AMCapability;

public class MagicBookScreen extends Screen {
    private static final int imgWidth = 512;
    private static final int imgHeight = 256;

    private static final ResourceLocation TEXTURE = textureGen("main");

    public MagicBookScreen() {
        super(Component.empty());
    }

    @Override
    public void render(@NotNull PoseStack pPoseStack, int xMouse, int yMouse, float partialTick) {
        super.render(pPoseStack, xMouse, yMouse, partialTick);

        int j = ((this.width / 2) - (imgWidth / 2));
        int k = ((this.height / 2) - (imgHeight / 2));

        assert this.minecraft != null;
        var player = this.minecraft.player;

        assert player != null;

        this.renderBackground(pPoseStack);
        RenderSystem.setShaderTexture(0, TEXTURE);

//        for (int i = 0; i < j; i++) {
//            for (int l = 0; l < k; l++) {
//                if (i == 66 && l == 29)
//                    RenderSystem.setShaderTexture(1, textureGen("player_display"));
//                if (i == 62 && l == 119)
//                    RenderSystem.setShaderTexture(2, textureGen("tag_display"));
//            }
//        }

        blit(pPoseStack, j, k, 0, 0, imgWidth, imgHeight, imgWidth, imgHeight);

        int xPos = j+97;
        int yPos = k+111;

        InventoryScreen.renderEntityInInventoryFollowsMouse(
                pPoseStack, xPos, yPos, 35, (float) xPos - xMouse, (float) yPos - yMouse,
                player
        );


        this.minecraft.font.draw(pPoseStack, Component.translatable("magic.ancientmagic.level",
                        MagicTypes.getByNumeration(ClientPlayerMagicData.getPlayerData()).getTranslation()),
                j + 56, k + 122, ChatFormatting.BLACK.getId());
    }

    private static ResourceLocation textureGen(String name) {
        return new ResourceLocation(Constant.Key, String.format("textures/screen/magicbook/%s.png", name));
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }
}

File: src/main/java/team/zeds/ancientmagic/api/screen/MagicBookScreen.kt
package team.zeds.ancientmagic.api.screen

import net.minecraft.ChatFormatting
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.gui.screens.inventory.InventoryScreen
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import team.zeds.ancientmagic.api.magic.MagicTypes
import team.zeds.ancientmagic.api.mod.Constant
import team.zeds.ancientmagic.client.packet.ClientPlayerMagicData

abstract class MagicBookScreen(component: Component): Screen(component), IMagicBookScreen {
    private val imgWidth: Int = 436
    private val imgHeight: Int = 232

    @Deprecated("Don't use MC method! Use from MagicBook.",
        ReplaceWith("render", "net.minecraft.client.gui.GuiGraphics",
        "kotlin.Int", "kotlin.Int", "kotlin.Float", "kotlin.Int", "kotlin.Int"),
        DeprecationLevel.HIDDEN
    )
    override fun render(guiGraphics: GuiGraphics, xMouse: Int, yMouse: Int, partialTick: Float) {
        super.render(guiGraphics, xMouse, yMouse, partialTick)
        val w = ((this.width / 2) - (imgWidth / 2))
        val h = ((this.height / 2) - (imgHeight / 2))

        val playerDisplayWidth = 64
        val playerDisplayHeight = 88
        val tagDisplayWidth = 92
        val tagDisplayHeight = 16

        if (this.getBookClassifier() == IMagicBookScreen.MagicBookClassifier.MAIN_PAGE) {
            guiGraphics.blit(textureGen("book_screen"), w, h, 0f, 0f, imgWidth, imgHeight, imgWidth, imgHeight)
            guiGraphics.blit(
                textureGen("player_frame"), w + 66, h + 29, 0f, 0f, playerDisplayWidth, playerDisplayHeight,
                playerDisplayWidth, playerDisplayHeight
            )
            guiGraphics.blit(
                textureGen("tag_frame"), w + 59, h + 119, 0f, 0f, tagDisplayWidth, tagDisplayHeight,
                tagDisplayWidth, tagDisplayHeight
            )
            val player = this.minecraft!!.player!!
            val xPos: Int = w + 97
            val yPos: Int = h + 111

            InventoryScreen.renderEntityInInventoryFollowsMouse(
                guiGraphics, xPos, yPos, 35, (xPos - xMouse).toFloat(),
                (yPos - yMouse).toFloat(), player
            )
            guiGraphics.drawString(font, Component.translatable(
                "magic.${Constant.KEY}.level", MagicTypes.getByNumeration(ClientPlayerMagicData.getPlayerData()).getTranslation()
            ),
                w + 56, h + 122, ChatFormatting.WHITE.id)
            render(guiGraphics, xMouse, yMouse, partialTick, w, h)
        } else render(guiGraphics, xMouse, yMouse, partialTick, w, h)
        super.render(guiGraphics, 0, 0, partialTick)
    }

    fun render(guiGraphics: GuiGraphics, xMouse: Int, yMouse: Int, partialTick: Float, width: Int, height: Int) {}

    fun textureGen(name: String): ResourceLocation {
        return ResourceLocation(Constant.KEY, "textures/screen/magicbook/${name}.png")
    }

    override fun isPauseScreen(): Boolean = false
}
